### PR TITLE
Llama index quickstart

### PR DESCRIPTION
**Pull request by michael-berk**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/michael-berk/mlflow/pull/12747?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12747/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12747
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
* Llama index quickstart notebook
* Llama index quickstart guide
* Move the FAQ to the llama index guide

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by michael-berk** [comment]
@mlflow-automation autoformat

**Comment by michael-berk** [comment]
@B-Step62 review callouts
* The guide is light relative to other MLflow flavors (langchain for example). I think is is inline with what we DMed about, but happy to add more material.
* I render the MLflow UI in an iframe in the quickstart. It doesn't render in the docs but does when you run the cell.

File: docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+    "\n",
+    "from llama_index.core import Document, VectorStoreIndex\n",
+    "from llama_index.core.llms import ChatMessage\n",
+    "from pyspark.sql import SparkSession\n",
+    "\n",
+    "import mlflow\n",
+    "\n",
+    "os.environ[\n",
+    "    \"OPENAI_API_KEY\"\n",
+    "] = \"sdk-...\"  # NOTE: this is for demo purposes - plaintext tokens are not secure!"
```
[the remark this answers is not in the document]
Comment (michael-berk) [reply docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb:77]: I don't think this works in a jupyter environment. We can do something like a widget but that persists after text is input. Suggestions?\

[the remark this answers is not in the document]
Comment (B-Step62) [reply docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb:77]: Screenshot 2024-07-25 at 8 36 05

Oh? It should work in notebook

[the remark this answers is not in the document]
Comment (michael-berk) [reply docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb:77]: Ok will add it but it's not rendering on my notebook. The notebook cell just runs and doesn't provide a way to populate the input.

[the remark this answers is not in the document]
Comment (B-Step62) [reply docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb]: Perhaps are you running the notebook in IDE? In VS code the prompt pops up in the command box on the top, rather then in the notebook.

File: docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Unique identifier for the model location for loading: runs:/036936a7ac964f0cb6ab99fa908d6421/llama_index\n"
+     ]
+    }
+   ],
+   "source": [
+    "mlflow.llama_index.autolog()\n",
```
Comment (B-Step62) [inline docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb:224]: nit: Can we add a single line comment saying this is for enabling trace? (not for logging model)

File: docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+    }
+   ],
+   "source": [
+    "print(\"\\n------------- Inference via Llama Index   -------------\")\n",
+    "index = mlflow.llama_index.load_model(model_uri)\n",
+    "query_response = index.as_query_engine().query(\"hi\")\n",
+    "print(query_response)\n",
+    "\n",
+    "print(\"\\n------------- Inference via MLflow PyFunc -------------\")\n",
+    "index = mlflow.pyfunc.load_model(model_uri)\n",
```
[the remark this answers is not in the document]
Comment (B-Step62) [reply docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb:298]: Can we fix this^?

**Review by B-Step62 — APPROVED** [review]
LGTM once the bullet points rendering are fixed and a few remaining comments are addressed! Awesome work!!

File: docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (michael-berk) [reply docs/source/llms/llama-index/notebooks/llama_index_quickstart.ipynb]: Yup great spot! My mistake

**Comment by michael-berk** [comment]
@mlflow-automation autoformat

**Pull request merged**